populate gossipsub scores

sinkingsugar opened the pull request:
inspired by LH scores

sinkingsugar commented:
From https://gist.github.com/blacktemplar/5c1862cb3f0e32a1a7fb0b25e79e6e2c#file-generate-scoring-params-py
Assumes around 50k validators, but looks like it can be realistic at this point...

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -365,10 +370,34 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
   var attestationSubscriptions: seq[Future[void]] = @[]
 
+  const topicParams = TopicParams(
+    topicWeight: 0.015625,
```
A general point is that where these come from isn't documented in the code. They have an arbitrary look at times, without such description. Even if/when the answer is, that's what LH does, that would be worth noting.

sinkingsugar replied:
Agreed, I'll add more rationale.

Are you comfortable with those consts inside proc bodies btw? Or you'd rather see them outside on top of the file?

I kinda like the contextual position but open to any opinion.

tersec replied:
When it's these single-use things, I agree with you and am fine with them in the procs, as-is. I usually do the same thing, only factoring them out once they have a second user. It minimizes namespace pollution and improves error messages by not giving the compiler more dead-ends to consider potentially relevant.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -449,15 +478,66 @@ proc getAttestationHandlers(node: BeaconNode): Future[void] =
+  node.network.subscribe(getAggregateAndProofsTopic(node.forkDigest), topicParams)
 
 proc addMessageHandlers(node: BeaconNode): Future[void] =
+  const blocksTopicParams = TopicParams(
```
The usual nbc style is to group
```
const a
const b
```
as
```
const
  a
  b
```

sinkingsugar committed:
force pushing to fix unstable base

sinkingsugar committed:
Merge branch 'unstable' into gossipsub-scores-full

arnetheduck committed:
increase attestation/aggregate queue sizes

when there are many validators, many aggregates and attestations arrive
every slot - increase the queue size a bit - also do batches on each
idle loop iteration since it's fairly quick

sinkingsugar committed:
Merge branch 'unstable' into gossipsub-scores-full

sinkingsugar committed:
Merge remote-tracking branch 'origin/longer-queues' into gossipsub-scores-full

sinkingsugar committed:
don't score subnets for now

sinkingsugar committed:
Merge remote-tracking branch 'origin/unstable' into gossipsub-scores-full

sinkingsugar commented:
Remaining to do:
A bit more local testing with new merged unstable
Add more comments

sinkingsugar committed:
wrapping up

sinkingsugar commented:
@tersec just pinged you for review but feel free to add more people if you think necessary.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -368,10 +373,15 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
     doAssert finalizedEpoch + 4 >= epoch
 
+proc getTotalValidators(node: BeaconNode): uint64 =
+  let epochRef = getEpochRef(node.chainDag, node.chainDag.head, node.chainDag.head.slot.epoch)
```
https://github.com/status-im/nimbus-eth2/blob/596b8c6009ce893bbafcbefc74bc3edeecf967ce/beacon_chain/block_pools/spec_cache.nim#L19-L21

sinkingsugar replied:
Ah I completely forgot about this.
It's not used yet. I might just take your comment as a reminder for the possible future and remove this for now.

tersec replied:
Yeah, that seems even more preferable.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -368,10 +373,15 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
+
 proc installAttestationSubnetHandlers(node: BeaconNode, subnets: set[uint8]) =
   # https://github.com/ethereum/eth2.0-specs/blob/v1.0.0/specs/phase0/p2p-interface.md#attestations-and-aggregation
+  # nimbus won't score attestation subnets for now, we just rely on block and aggregate which are more stabe and reliable
```
It looks like attester slashings, proposer slashings, and voluntary exits get `basicParams`, which is presumably not the same as not scoring them?

tersec replied:
Oh, they are the same.

sinkingsugar replied:
Yep, I tested for a whole and honestly just scoring blocks and aggregates is plenty.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -671,11 +681,60 @@ proc getAttestationSubnetHandlers(node: BeaconNode) =
+      meshMessageDeliveriesCap: 2.054757422716966,
+      meshMessageDeliveriesActivation: chronos.seconds(384),
+      meshMessageDeliveriesWindow: chronos.seconds(2),
+      meshFailurePenaltyWeight: -458.31054878249114 ,
```
This and the other prioritization of blocks over aggregates seems appropriate.

tersec commented on beacon_chain/eth2_network.nim:
```diff
@@ -1567,7 +1567,20 @@ proc createEth2Node*(rng: ref BrHmacDrbgContext,
+        p.gossipThreshold = -4000
+        p.publishThreshold = -8000
+        p.graylistThreshold = -16000
+        p.decayInterval = 12.seconds
```
Does `GossipSubParams` have an `init` more along the lines of `Foo(d: 8, dLow: 6, ...)` available? This is really just one single object construction, which seems better not to do as a couple dozen assignment statements.

sinkingsugar replied:
Yeah good point.

sinkingsugar committed:
refactor and cleanups

sinkingsugar committed:
Merge remote-tracking branch 'origin/unstable' into gossipsub-scores-full

sinkingsugar committed:
gossip parameters fixes

tersec commented on beacon_chain/eth2_network.nim:
```diff
@@ -1563,7 +1563,7 @@ proc createEth2Node*(rng: ref BrHmacDrbgContext,
       dHigh: 12,
       dScore: 6,
-      dOut: 6 - 1, # DLow - 1
+      dOut: 6 div 2, # DLow - 1
```
6 div 2 isn't DLow - 1?

sinkingsugar replied:
comment is incorrect yeah it has to be either less than dlow but not more then dlow/2

sinkingsugar committed:
comment fix
